pkg/kubefed2/federate/federate.go

```diff
@@ -0,0 +1,281 @@
+			if err != nil {
+				glog.Fatalf("error: %v", err)
+			}
+		},
+	}
+
+	flags := cmd.Flags()
+	opts.CommonBind(flags)
+	opts.Bind(flags)
+
+	return cmd
+}
+
+// Run is the implementation of the `federate resource` command.
+func (j *federateResource) Run(cmdOut io.Writer, config util.FedConfig) error {
+	hostConfig, err := config.HostConfig(j.HostClusterContext, j.Kubeconfig)
+	if err != nil {
+		return errors.Wrap(err, "Failed to get host cluster config")
+	}
+
+	qualifiedTypeName := ctlutil.QualifiedName{
+		Namespace: j.FederationNamespace,
+		Name:      j.typeName,
+	}
+	qualifiedResourceName := ctlutil.QualifiedName{
+		Namespace: j.resourceNameSpace,
+		Name:      j.resourceName,
+	}
+
+	return FederateResource(hostConfig, qualifiedTypeName, qualifiedResourceName)
+}
+
+func FederateResource(hostConfig *rest.Config, qualifiedTypeName, qualifiedResourceName ctlutil.QualifiedName) error {
+	typeConfig, err := lookUpTypeDetails(hostConfig, qualifiedTypeName)
+	if err != nil {
+		return err
+	}
+
+	templateResource, err := getTargetResource(hostConfig, typeConfig, qualifiedResourceName)
+	if err != nil {
+		return err
+	}
+
+	return createFedResource(hostConfig, typeConfig, templateResource)
+}
+
+func lookUpTypeDetails(config *rest.Config, qualifiedTypeName ctlutil.QualifiedName) (*fedv1a1.FederatedTypeConfig, error) {
+	fedClient, err := util.FedClientset(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to get federation clientset")
+	}
+
+	typeConfig, err := fedClient.CoreV1alpha1().FederatedTypeConfigs(qualifiedTypeName.Namespace).Get(qualifiedTypeName.Name, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, errors.Wrapf(err, "FederatedTypeConfig %q, does not exist. Type needs to be enabled first", qualifiedTypeName)
+		}
+		return nil, errors.Wrapf(err, "Error retrieving FederatedTypeConfig %q", qualifiedTypeName)
+	}
+
+	crdClient, err := apiextv1b1client.NewForConfig(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error creating crd client")
+	}
+
+	_, err = crdClient.CustomResourceDefinitions().Get(typeconfig.GroupQualifiedName(typeConfig.GetFederatedType()), metav1.GetOptions{})
+	if err != nil {
+		return nil, errors.Wrapf(err, "Error retrieving Federated<type> CRD corresponding to FederatedTypeConfig %q. Federated<type> CRD should exist", qualifiedTypeName)
+	}
+
+	return typeConfig, err
+}
+
+func getTargetResource(hostConfig *rest.Config, typeConfig *fedv1a1.FederatedTypeConfig, qualifiedResourceName ctlutil.QualifiedName) (*unstructured.Unstructured, error) {
+	targetAPIResource := typeConfig.GetTarget()
+	targetClient, err := ctlutil.NewResourceClient(hostConfig, &targetAPIResource)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error creating target resource client")
+	}
+
+	resource, err := targetClient.Resources(qualifiedResourceName.Namespace).Get(qualifiedResourceName.Name, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
```

Comment:
What's the value of treating 'not found' differently from any other error in this instance?

Reply:
`Not found` signifies naming issue or resource genuinely missing; Differentiating it with say a permission or an access error makes sense.

Reply:
Does the error message returned by the client not provide sufficient differentiation?